### PR TITLE
fix: exclude unstable data from cache fingerprints

### LINKED ISSUE #18116: Preset changes cause unnecessary cache busting
### Describe the proposed change(s).

The hosted app was upgraded between these versions: https://app.renovatebot.com/package-diff?name=renovate&from=32.214.0&to=32.216.0

Clone percentage immediately went up to ~100%

Why?

Random example from public repo https://github.com/larvit/larvitsession which had no commit for 9 days:

Last job log before upgrade:

```
{"level":20,"msg":"Using app token for git init","time":"2022-10-05T03:09:46.830Z"}
{"level":20,"msg":"Repository cache is restored from revision 13","time":"2022-10-05T03:09:47.238Z"}
{"level":20,"msg":"Resetting npmrc","time":"2022-10-05T03:09:47.238Z"}
{"level":20,"msg":"detectSemanticCommits()","time":"2022-10-05T03:09:47.239Z"}
{"level":20,"msg":"checkOnboarding()","time":"2022-10-05T03:09:47.239Z"}
{"level":20,"msg":"isOnboarded()","time":"2022-10-05T03:09:47.239Z"}
{"level":20,"msg":"Checking cached config file name","time":"2022-10-05T03:09:47.239Z"}
{"level":20,"msg":"Existing config file confirmed","time":"2022-10-05T03:09:47.474Z"}
{"level":20,"fileName":"renovate.json","config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>larvit/standards:renovate-default"]},"msg":"Repository config","time":"2022-10-05T03:09:47.474Z"}
{"level":20,"msg":"Repo is onboarded","time":"2022-10-05T03:09:47.474Z"}
{"level":20,"msg":"migrateAndValidate()","time":"2022-10-05T03:09:47.475Z"}
{"level":20,"msg":"No config migration necessary","time":"2022-10-05T03:09:47.476Z"}
{"level":20,"config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>whitesource/merge-confidence:beta","github>larvit/standards:renovate-default"]},"msg":"massaged config","time":"2022-10-05T03:09:47.476Z"}
{"level":20,"config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>whitesource/merge-confidence:beta","github>larvit/standards:renovate-default"]},"msg":"migrated config","time":"2022-10-05T03:09:47.476Z"}
{"level":20,"msg":"Setting hostRules from config","time":"2022-10-05T03:09:48.721Z"}
{"level":20,"ignorePaths":["**/node_modules/**","**/bower_components/**","**/vendor/**","**/examples/**","**/__tests__/**","**/test/**","**/tests/**","**/__fixtures__/**"],"msg":"Found repo ignorePaths","time":"2022-10-05T03:09:48.768Z"}
{"level":20,"concurrency":10,"host":"api.github.com","msg":"Using queue","time":"2022-10-05T03:09:48.772Z"}
{"level":20,"msg":"No vulnerability alerts found","time":"2022-10-05T03:09:48.990Z"}
{"level":20,"msg":"No vulnerability alerts found","time":"2022-10-05T03:09:48.990Z"}
{"level":20,"msg":"findIssue(Dependency Dashboard)","time":"2022-10-05T03:09:48.991Z"}
{"level":20,"msg":"Retrieving issueList","time":"2022-10-05T03:09:48.991Z"}
{"level":20,"msg":"Retrieved 1 issues","time":"2022-10-05T03:09:49.215Z"}
{"level":20,"msg":"Found issue 49","time":"2022-10-05T03:09:49.215Z"}
{"level":20,"msg":"No baseBranches","time":"2022-10-05T03:09:49.395Z"}
{"level":20,"msg":"extract()","time":"2022-10-05T03:09:49.396Z"}
{"level":20,"baseBranch":"master","baseBranchSha":"b8619f8dc5b55bd65bddfc64f220ef69e09d9c97","msg":"Found cached extract","time":"2022-10-05T03:09:49.397Z"}
```

After upgrade:

```
{"level":20,"msg":"Using app token for git init","time":"2022-10-05T05:39:37.737Z"}
{"level":20,"msg":"Repository cache is restored from revision 13","time":"2022-10-05T05:39:38.155Z"}
{"level":20,"msg":"Resetting npmrc","time":"2022-10-05T05:39:38.155Z"}
{"level":20,"msg":"detectSemanticCommits()","time":"2022-10-05T05:39:38.156Z"}
{"level":20,"msg":"checkOnboarding()","time":"2022-10-05T05:39:38.156Z"}
{"level":20,"msg":"isOnboarded()","time":"2022-10-05T05:39:38.156Z"}
{"level":20,"msg":"Checking cached config file name","time":"2022-10-05T05:39:38.156Z"}
{"level":20,"msg":"Existing config file confirmed","time":"2022-10-05T05:39:38.394Z"}
{"level":20,"fileName":"renovate.json","config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>larvit/standards:renovate-default"]},"msg":"Repository config","time":"2022-10-05T05:39:38.395Z"}
{"level":20,"msg":"Repo is onboarded","time":"2022-10-05T05:39:38.395Z"}
{"level":20,"msg":"migrateAndValidate()","time":"2022-10-05T05:39:38.396Z"}
{"level":20,"msg":"No config migration necessary","time":"2022-10-05T05:39:38.397Z"}
{"level":20,"config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>whitesource/merge-confidence:beta","github>larvit/standards:renovate-default"]},"msg":"massaged config","time":"2022-10-05T05:39:38.397Z"}
{"level":20,"config":{"$schema":"https://docs.renovatebot.com/renovate-schema.json","extends":["github>whitesource/merge-confidence:beta","github>larvit/standards:renovate-default"]},"msg":"migrated config","time":"2022-10-05T05:39:38.397Z"}
{"level":20,"msg":"Setting hostRules from config","time":"2022-10-05T05:39:39.829Z"}
{"level":20,"ignorePaths":["**/node_modules/**","**/bower_components/**","**/vendor/**","**/examples/**","**/__tests__/**","**/test/**","**/tests/**","**/__fixtures__/**"],"msg":"Found repo ignorePaths","time":"2022-10-05T05:39:39.874Z"}
{"level":20,"concurrency":10,"host":"api.github.com","msg":"Using queue","time":"2022-10-05T05:39:39.879Z"}
{"level":20,"msg":"No vulnerability alerts found","time":"2022-10-05T05:39:40.099Z"}
{"level":20,"msg":"No vulnerability alerts found","time":"2022-10-05T05:39:40.099Z"}
{"level":20,"msg":"findIssue(Dependency Dashboard)","time":"2022-10-05T05:39:40.100Z"}
{"level":20,"msg":"Retrieving issueList","time":"2022-10-05T05:39:40.100Z"}
{"level":20,"msg":"Retrieved 1 issues","time":"2022-10-05T05:39:40.332Z"}
{"level":20,"msg":"Found issue 49","time":"2022-10-05T05:39:40.332Z"}
{"level":20,"msg":"No baseBranches","time":"2022-10-05T05:39:40.548Z"}
{"level":20,"msg":"extract()","time":"2022-10-05T05:39:40.548Z"}
{"level":20,"msg":"Setting current branch to master","time":"2022-10-05T05:39:40.549Z"}
{"level":20,"msg":"Initializing git repository into /mnt/renovate/gh/larvit/larvitsession","time":"2022-10-05T05:39:40.550Z"}
{"level":20,"msg":"Performing blobless clone","time":"2022-10-05T05:39:40.550Z"}
```

I don't see why the config change caused by preset addition should have cache-busted the extract.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Excludes `packageRules` config from the extract cache fingerprint so that our regular monorepo changes/improvements don't invalidate extract caches every time.

Also uses stable stringify so that ordering is not a problem in future.

## Context

Closes #18116

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/process/extract-update.ts [hunk 1]
```diff
@@ -71,7 +72,10 @@ export async function extract(
   const cache = getCache();
   cache.scan ||= {};
   const cachedExtract = cache.scan[baseBranch!];
-  const configHash = hasha(JSON.stringify(config));
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { packageRules, ...remainingConfig } = config;
```
Comment (rarkins) [inline lib/workers/repository/process/extract-update.ts:76]: I didn't want to clone it unnecessarily (so I could delete the field), couldn't think of a way to do this without the eslint rule being triggered

File: lib/workers/repository/process/extract-update.spec.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import hasha from 'hasha';
+import stringify from 'safe-stable-stringify';
```
Comment (viceice) [inline lib/workers/repository/process/extract-update.spec.ts:2]: we've a util function for this which should be used

File: lib/workers/repository/process/extract-update.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // TODO #7154
 import is from '@sindresorhus/is';
 import hasha from 'hasha';
+import stringify from 'safe-stable-stringify';
```
Comment (viceice) [inline lib/workers/repository/process/extract-update.ts:4]: see other comment

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/workers/repository/process/extract-update.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/process/extract-update.ts:75]: Should we change `ignoreRestSiblings` to true in eslint rules? Currently it is false. 
[docs](https://eslint.org/docs/latest/rules/no-unused-vars#ignorerestsiblings)

Comment (RahulGautamSingh) [reply lib/workers/repository/process/extract-update.ts:75]: Should we change ignoreRestSiblings to true in eslint rules? Currently it is false.
[docs](https://eslint.org/docs/latest/rules/no-unused-vars#ignorerestsiblings)

File: lib/config/options/index.ts [hunk 4]
```diff
@@ -382,6 +382,7 @@ const options: RenovateOptions[] = [
     globalOnly: true,
     type: 'string',
     default: null,
+    stage: 'global',
```
Comment (rarkins) [inline lib/config/options/index.ts:385]: This makes sure that random `logContext` values don't stay in the config and cachebust

File: lib/workers/repository/process/extract-update.ts (hunk 1 above)
Comment (rarkins) [reply lib/workers/repository/process/extract-update.ts:75]: Good idea

File: lib/util/fingerprint.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import hasha from 'hasha';
+import { safeStringify } from './stringify';
+
+export function fingerprint(input: unknown): string {
+  return hasha(safeStringify(input), { algorithm: 'sha256' });
```
Comment (viceice) [inline lib/util/fingerprint.ts:5]: why now back to sha256 instead of default sha512 like before?

**Review by viceice — APPROVED** [review]

**Pull request merged**